sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,517 @@
+          i += 2
+        case c: Char =>
+          tokens :+= InvalidUnrecognizedCharacter(c)
+          i += 1
+      }
+    }
+
+    // Combine each group of consecutive Digits and ThousandsSeparator tokens into a DigitGroups.
+    i = 0
+    var groupedTokens = mutable.Seq.empty[InputToken]
+    while (i < tokens.length) {
+      val suffix: mutable.Seq[InputToken] = tokens.drop(i)
+      val gatheredTokens: mutable.Seq[InputToken] = suffix.takeWhile {
+        case _: Digits | _: ThousandsSeparator => true
+        case _ => false
+      }
+      if (gatheredTokens.nonEmpty) {
+        groupedTokens ++= mutable.Seq[InputToken](DigitGroups(gatheredTokens.reverse))
+        i += gatheredTokens.length
+      } else {
+        groupedTokens :+= tokens(i)
+        i += 1
+      }
+    }
+    groupedTokens
+  }
+
+  /**
+   * Precision is the number of digits in a number. Scale is the number of digits to the right of
+   * the decimal point in a number. For example, the number 123.45 has a precision of 5 and a
+   * scale of 2.
+   */
+  private lazy val precision = {
+    val lengths = inputTokens.map {
+      case DigitGroups(tokens) => tokens.map {
+        case ExactlyAsManyDigits(num) => num
+        case AtMostAsManyDigits(num) => num
+        case _: ThousandsSeparator => 0
+      }.sum
+      case _ => 0
+    }
+    lengths.sum
+  }
+
+  private lazy val scale = {
+    val index = inputTokens.indexOf(DecimalPoint())
+    if (index != -1) {
+      val decimalPointIndex = inputTokens.indexOf(DecimalPoint())
+      val suffix: Seq[InputToken] = inputTokens.drop(decimalPointIndex)
+      val lengths: Seq[Int] = suffix.map {
+        case DigitGroups(tokens) => tokens.map {
+          case ExactlyAsManyDigits(num) => num
+          case AtMostAsManyDigits(num) => num
+        }.sum
+        case _ => 0
+      }
+      lengths.sum
+    } else {
+      0
+    }
+  }
+
+  /**
+   * The result type of this parsing is a Decimal value with the appropriate precision and scale.
+   */
+  def parsedDecimalType: DecimalType = DecimalType(precision, scale)
+
+  /**
+   * Consumes the format string to check validity and computes an appropriate Decimal output type.
+   */
+  def check(): TypeCheckResult = {
+    val validateResult: String = validateFormatString
+    if (validateResult.nonEmpty) {
+      TypeCheckResult.TypeCheckFailure(validateResult)
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  /**
+   * This implementation of the [[check]] method returns any error, or the empty string on success.
+   */
+  private def validateFormatString: String = {
+    def multipleSignInNumberFormatError(message: String) = {
+      s"At most one $message is allowed in the number format: '$originNumberFormat'"
+    }
+
+    def notAtEndOfNumberFormatError(message: String) = {
+      s"$message must be at the end of the number format: '$originNumberFormat'"
+    }
+
+    def inputTokenCounts = inputTokens.groupBy(identity).mapValues(_.size)
```

Comment:
this should be a `val`, otherwise we do the `groupBy` many times.

Reply:
Done.